Enforce immediate conflict resolution

Larisa-Staroverova opened the pull request.

Larisa-Staroverova committed:
feat: Variant carousel (#15914, #15911, #15988)

Enable to display variant carousel for configurable products

Closes #15894, #15895, #15896

Larisa-Staroverova committed:
Merge branch 'develop' into epic/configuratorsAfter50

Larisa-Staroverova committed:
Define new conflict-solver-dialog component

Larisa-Staroverova committed:
Fix import

Larisa-Staroverova committed:
Merge branch 'develop' into epic/configuratorsAfter50

Larisa-Staroverova committed:
Merge branch 'epic/configuratorsAfter50' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Build in discussed changes

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Replace modalService through launchDialogService

Larisa-Staroverova committed:
Refactoring

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Refactoring styling, add tests

Larisa-Staroverova committed:
Trigger Build

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

github-actions[bot] committed:
Prepend licence to files

Larisa-Staroverova committed:
Add routing service

Larisa-Staroverova committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Refactoring: after talk with Anthony

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Uli-Tiger committed:
feature: make conflict to group navigation optional and hide it in conflict solver dialog [CXSPA-1865] (#16655)

Larisa-Staroverova committed:
Enhance test coverage

Uli-Tiger committed:
fix: do not open modal when navigating to overview page

Larisa-Staroverova committed:
Rename dialog layout file

Larisa-Staroverova committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Make prettier happy

Larisa-Staroverova committed:
Enhance styling

Uli-Tiger committed:
fix: empty conflict modal on ov navigation

Uli-Tiger committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

Uli-Tiger committed:
test; use hot observable

Uli-Tiger committed:
pretty print marble tests

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Enhance styling

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog.component.ts:
```diff
@@ -0,0 +1,75 @@
+    this.routerData$ = routerData;
+  }
+
+  updateConfiguration(event: ConfigFormUpdateEvent): void {
+    this.configuratorCommonsService.updateConfiguration(
+      event.ownerKey,
+      event.changedAttribute,
+      event.updateType
+    );
```
can plz. provide JSDoc for this public declaration

Larisa-Staroverova replied:
Done!

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog.component.ts:
```diff
@@ -0,0 +1,75 @@
+    this.subscription?.unsubscribe();
+  }
+
+  dismissModal(reason?: any): void {
+    this.launchDialogService.closeDialog(reason);
+  }
```
can plz. provide JSDoc for this public declaration

Larisa-Staroverova replied:
Done!

Uli-Tiger commented on feature-libs/product-configurator/rulebased/core/facade/configurator-groups.service.ts:
```diff
@@ -168,6 +168,20 @@ export class ConfiguratorGroupsService {
     );
   }
 
+  getConflictGroups(
+    owner: CommonConfigurator.Owner
+  ): Observable<Configurator.Group[] | undefined> {
+    return this.configuratorCommonsService
+      .getConfiguration(owner)
+      .pipe(
+        map((configuration) =>
+          configuration.flatGroups.filter(
+            (group) => group.groupType === Configurator.GroupType.CONFLICT_GROUP
+          )
+        )
+      );
+  }
+
```
can plz. provide JSDoc for this public declaration

Larisa-Staroverova replied:
Done!

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Uli-Tiger commented:
the overlay is misplaced in desktop resolution, so that the close button is inaccessible
![image](https://user-images.githubusercontent.com/53090888/208908230-3ba05583-4aff-4d28-97c8-62bbf76b0527.png)

Uli-Tiger commented:
to be consistent with the filter modal, the conflict modal shall be closed when clicking outside, or when pressing ESC-Key.

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog-event.listener.ts:
```diff
@@ -0,0 +1,107 @@
+      dialogData
+    );
+
+    if (dialog) {
+      dialog.pipe(take(1)).subscribe();
+    }
```
instead of subscribing yourself, one could also use `this.launchDialogService.openDialogAndSubscribe()`
this would ease the coding a bit

Larisa-Staroverova replied:
Agree! Done!

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog.component.html:
```diff
@@ -0,0 +1,176 @@
+    <div class="cx-modal-footer">
+      <!--TODO: here comes 'DONE' button-->
+    </div>
+  </div>
```
i can delete the entire HTML - still the component test is green?

Larisa-Staroverova replied:
Agree! Working at it!

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog.component.spec.ts:
```diff
@@ -0,0 +1,127 @@
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
```
as indicated above some components tests are missing, so for example one could assert
- if we have a configuration without conflicts the modal is not rendered
- if we have a configuration with one conflict, the modal is rendered and the conflict is visualized using the expected components depending on UI Type (ideally each UI Type is covered)
-  if we have a configuration with more than one conflict, only the first one is visualized

Larisa-Staroverova replied:
@Uli-Tiger, the first point is not valid according to my point of view because the launching of the modal depends on the dialog event listener. The listener will not open the modal if there are no conflicts, that's why it is overengineered to write a test for it.

Uli-Tiger replied:
agree

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog.component.html:
```diff
@@ -0,0 +1,176 @@
+                (selectionChange)="updateConfiguration($event)"
+              ></cx-configurator-attribute-radio-button>
+
+              <!-- Single Selection Product Bundle -->
+              <cx-configurator-attribute-single-selection-bundle
+                (selectionChange)="updateConfiguration($event)"
+                [attribute]="attribute"
+                [ownerKey]="routerData.owner.key"
+                *ngIf="attribute.uiType === uiType.RADIOBUTTON_PRODUCT"
+              ></cx-configurator-attribute-single-selection-bundle>
+
+              <!-- Multi Selection Product Bundle -->
+              <cx-configurator-attribute-multi-selection-bundle
+                (selectionChange)="updateConfiguration($event)"
+                [attribute]="attribute"
+                [ownerKey]="routerData.owner.key"
+                *ngIf="attribute.uiType === uiType.CHECKBOXLIST_PRODUCT"
+              ></cx-configurator-attribute-multi-selection-bundle>
+
+              <!-- Single Selection Dropdown Product Bundle -->
+              <cx-configurator-attribute-single-selection-bundle-dropdown
+                *ngIf="attribute.uiType === uiType.DROPDOWN_PRODUCT"
+                [attribute]="attribute"
+                [group]="conflictGroups[0].id"
+                [ownerKey]="routerData.owner.key"
+                (selectionChange)="updateConfiguration($event)"
+              ></cx-configurator-attribute-single-selection-bundle-dropdown>
```
i believe we can remove those, because bundles are only supported by CPQ and not by AVC.

Larisa-Staroverova replied:
Agree!

Larisa-Staroverova replied:
Done!

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog.component.html:
```diff
@@ -0,0 +1,176 @@
+      </ng-container>
+    </div>
+
+    <!-- Modal Footer -->
+    <div class="cx-modal-footer">
+      <!--TODO: here comes 'DONE' button-->
+    </div>
```
plz remove empty footer and TODO

Larisa-Staroverova replied:
Agree! Done!

Larisa-Staroverova replied:
Done!

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog.component.spec.ts:
```diff
@@ -0,0 +1,127 @@
+      const reason = 'Close conflict solver dialog';
+      component.ngOnInit();
+      component.dismissModal(reason);
+      expect(launchDialogService.closeDialog).toHaveBeenCalled();
+      expect(launchDialogService.closeDialog).toHaveBeenCalledWith(reason);
```
i think expect in line 124 is sufficient so that line 123 can be deleted

Larisa-Staroverova replied:
Done!

Larisa-Staroverova replied:
Done!

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog.module.ts:
```diff
@@ -0,0 +1,67 @@
+    ConfiguratorAttributeMultiSelectionImageModule,
+    ConfiguratorConflictDescriptionModule,
+    ConfiguratorConflictSuggestionModule,
+    ConfiguratorExitButtonModule,
```
is the ConfiguratorExitButtonModule really required?

Larisa-Staroverova replied:
You are right! I will remove it! Done!

Larisa-Staroverova replied:
Removed!

Uli-Tiger commented:
in VS code i get the following typescript error
![image](https://user-images.githubusercontent.com/53090888/208932701-00f07bcb-a6a5-4b3c-acfc-1c68ad82b333.png)

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Review feedback

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Enhance test coverage

ChristophHi reviewed (changes requested):
It seems that the PR does not check yet on configuration attribute immediateConflictResolution from OCC

ChristophHi commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog-event.listener.ts:
```diff
@@ -0,0 +1,100 @@
+@Injectable({
+  providedIn: 'root',
+})
+export class ConfiguratorConflictSolverDialogEventListener
```
This class is not listening to any event, what about renaming to ConfiguratorConflictDialogLauncherService?
@Larisa-Staroverova , @Uli-Tiger What do you think of removing 'solver' from the names of the entities in order to not make them that long? It's clear that we want to solve conflicts. Naming of the default config needs to be considered specifically.

Larisa-Staroverova replied:
There is a `ClearCartDialogComponentService` service as an example that injects`LaunchDialogService` to close the clear acrt dialog. That's why I would suggest to rename ours to `ConfiguratorConflictDialogComponentService`.

Larisa-Staroverova replied:
Done!

Uli-Tiger replied:
agree with Christoph that 'Solver' is not required in the names. So for example the launcher could be called `ConfiguratorConflictDialogLauncherService`

ChristophHi commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog-event.listener.spec.ts:
```diff
@@ -0,0 +1,228 @@
+      overviewRouterData.pageType = ConfiguratorRouter.PageType.OVERVIEW;
+    });
+    it('should emit group data when routing pageType = "configuration"', () => {
+      routerData$ = hot('     c---', { c: configRouterData });
```
Observables can all be 'cold' here (or all 'hot'). No need to have one cold and one hot.
For me it seems cold is more natural as it comes closer to how our actual observables behave

Uli-Tiger replied:
done

ChristophHi commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/index.ts:
```diff
@@ -0,0 +1,8 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+export * from './configurator-conflict-solver-dialog.component';
```
Please also export the event listener (or whatever we call this class)

Larisa-Staroverova replied:
Done!

ChristophHi commented on feature-libs/product-configurator/rulebased/core/facade/configurator-groups.service.ts:
```diff
@@ -168,6 +168,26 @@ export class ConfiguratorGroupsService {
+   */
+  getConflictGroups(
+    owner: CommonConfigurator.Owner
+  ): Observable<Configurator.Group[] | undefined> {
```
```suggestion
  ): Observable<Configurator.Group[] > {
```

Larisa-Staroverova replied:
Adjusted and test coverage has been enhanced.

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Add new default-form component

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Enhance test coverage for configurator-form

Larisa-Staroverova committed:
Enhance test coverage

Larisa-Staroverova committed:
Enhance test coverage for default-form

Larisa-Staroverova committed:
Review feedback

Larisa-Staroverova committed:
Review feedback

Larisa-Staroverova committed:
Enhance normalizer with immediateConflictResolution attribute

Larisa-Staroverova committed:
Remove resolve conflicts group, diable conflict detected link in module and in configuration form  for immediate conflict resolution

Larisa-Staroverova committed:
Enable tests

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Adjust previous button navigation for AVC configuration

Larisa-Staroverova committed:
Clone structure of test data

Larisa-Staroverova committed:
Make prettier happy

Larisa-Staroverova committed:
Remove import

Larisa-Staroverova committed:
Make sonar happy

Larisa-Staroverova committed:
Make sonar happy II

Larisa-Staroverova committed:
Make some makeovers

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

github-actions[bot] committed:
Add license header

Larisa-Staroverova committed:
Revert "Add license header"

This reverts commit c905bf3cc0489cc4948b4be6468dc697841327fd.

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

github-actions[bot] committed:
Add license header

Larisa-Staroverova committed:
Fix license header

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Fix issues with update of checkbox list

Uli-Tiger committed:
review: use only cold marbles

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog-launcher.service.ts:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  protected openConflictSolverDialog() {
+    this.subscription.add(
+      this.conflictGroups$.subscribe((conflictGroups) => {
+        if (conflictGroups && conflictGroups?.length > 0) {
+          this.openModal();
+        }
+      })
+    );
+
+    this.subscription.add(
+      this.conflictGroups$.subscribe((conflictGroups) => {
+        if (conflictGroups && conflictGroups?.length === 0) {
+          this.closeModal('CLOSE_NO_CONFLICTS_EXIST');
+        }
+      })
+    );
```
@Larisa-Staroverova: i believe both subscriptions could be merged into one
```suggestion
  this.subscription.add(
      this.conflictGroups$.subscribe((conflictGroups) => {
        if (conflictGroups && conflictGroups?.length > 0) {
          this.openModal();
        } else {
          this.closeModal('CLOSE_NO_CONFLICTS_EXIST');
        }
      })
    );
```

Larisa-Staroverova replied:
Good point! Done!

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog-launcher.service.ts:
```diff
@@ -0,0 +1,100 @@
+    return semanticRoute ? semanticRoute.includes('configure') : false;
+  }
+
+  protected openConflictSolverDialog() {
```
@Larisa-Staroverova: Can you rename this method as it is not directly opening the conflict solver dialog, but subscribing to conflict group and router observables in order to close/open the modal depending on router and conflict state.

Larisa-Staroverova replied:
Good idea!

Larisa-Staroverova replied:
@Uli-Tiger, do you have any idea? What about `conrolDialog`?

Uli-Tiger commented on feature-libs/product-configurator/rulebased/components/default-form/configurator-default-form.component.ts:
```diff
@@ -0,0 +1,139 @@
+    protected configExpertModeService: ConfiguratorExpertModeService
+  ) {}
+
+  ngOnInit(): void {
+    this.configRouterExtractorService
+      .extractRouterData()
+      .pipe(take(1))
+      .subscribe((routingData) => {
+        //In case of resolving issues, check if the configuration contains conflicts,
+        //if not, check if the configuration contains missing mandatory fields and show the group
+        if (routingData.resolveIssues) {
+          this.configuratorCommonsService
+            .hasConflicts(routingData.owner)
+            .pipe(take(1))
+            .subscribe((hasConflicts) => {
+              if (hasConflicts && !routingData.skipConflicts) {
+                this.configuratorGroupsService.navigateToConflictSolver(
+                  routingData.owner
+                );
+
+                //Only check for Incomplete group when there are no conflicts or conflicts should be skipped
+              } else {
+                this.configuratorGroupsService.navigateToFirstIncompleteGroup(
+                  routingData.owner
+                );
+              }
+            });
+        }
+        if (routingData.expMode) {
+          this.configExpertModeService?.setExpModeRequested(
+            routingData.expMode
+          );
+        }
+      });
+  }
```
@Larisa-Staroverova: i believe this code makes only sense in context of the form-component, but not the dialog-solver-component, so i suggest to move it to the form-component.

Larisa-Staroverova replied:
Agree!

Larisa-Staroverova replied:
Done!

Uli-Tiger commented:
> in VS code i get the following typescript error ![image](https://user-images.githubusercontent.com/53090888/208932701-00f07bcb-a6a5-4b3c-acfc-1c68ad82b333.png)

This is still the case, should't we fix it?

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Review feedback

ChristophHi committed:
experiment: store showConflictSolverDialogue in state

ChristophHi committed:
Merge branch 'feature/GH-15991' into feature/GH-15991-state

Larisa-Staroverova committed:
review feedback II

ChristophHi committed:
Call new facade method

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Add comment to the method

ChristophHi committed:
Resolve conflicts

Uli-Tiger commented:
lgtm

ChristophHi reviewed (changes requested):
Let's only merge when we figured out the best solution for not rendering attributes with the same Ids, also this needs to go into a major release. Moreover, we need to do another round of reviews and testing

ChristophHi committed:
group menu should be displayed on dialog close

ChristophHi commented on feature-libs/product-configurator/rulebased/components/config/default-configurator-ui-settings.config.ts:
```diff
@@ -14,6 +14,6 @@ export const defaultConfiguratorUISettingsConfig: ConfiguratorUISettingsConfig =
       },
       addRetractOption: false,
-      enableNavigationToConflict: false,
+      enableNavigationToConflict: true,
```
@Larisa-Staroverova Please revert this change

Larisa-Staroverova replied:
Done!

ChristophHi commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog-launcher.service.spec.ts:
```diff
@@ -0,0 +1,232 @@
+import {
+  Configurator,
+  ConfiguratorGroupsService,
+} from '@spartacus/product-configurator/rulebased';
```
Please import from relative location

Larisa-Staroverova replied:
Done!

ChristophHi commented on feature-libs/product-configurator/rulebased/components/conflict-solver-dialog/configurator-conflict-solver-dialog.component.spec.ts:
```diff
@@ -0,0 +1,218 @@
+  ConfigFormUpdateEvent,
+  Configurator,
+  ConfiguratorCommonsService,
+} from '@spartacus/product-configurator/rulebased';
```
Wrong import

Larisa-Staroverova replied:
Done!

ChristophHi commented on feature-libs/product-configurator/rulebased/components/default-form/index.ts:
```diff
@@ -0,0 +1,9 @@
+ */
+
+export * from './configurator-default-form.component';
+export * from './configurator-default-form.event';
```
Why do we have to move the event? Having it at the same location saves us many changes in this PR

Larisa-Staroverova replied:
For the sake of completeness, I will say. Both `form` and `conflict-solver-dialog` inject `default-form`. According to my point of view, it could be questionable if the event stays in `form` and maybe one could ask oneself why not in `conflict-solver-dialog`. We came up to the conclusion that the common part could be extracted from the form and the event should belong to this common part as well.

ChristophHi commented on feature-libs/product-configurator/rulebased/components/default-form/configurator-default-form.component.spec.ts:
```diff
@@ -0,0 +1,644 @@
+  ConfigFormUpdateEvent,
+  ConfiguratorConflictSuggestionComponent,
+  ConfiguratorStorefrontUtilsService,
+} from '@spartacus/product-configurator/rulebased';
```
wrong import

Larisa-Staroverova replied:
Done!

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Review feedback and fix imports in specs

Uli-Tiger committed:
fix: delete dead code

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Uli-Tiger commented:
for small screens in landscape mode the title is cut off - for example use 1000x600
![image](https://user-images.githubusercontent.com/53090888/211363401-808b2ac6-a16e-469c-a990-ac10549216e1.png)

Uli-Tiger reviewed (changes requested):
@Larisa-Staroverova: for small screens in landscape mode the title is cut off - for example use 1000x600
![image](https://user-images.githubusercontent.com/53090888/211363401-808b2ac6-a16e-469c-a990-ac10549216e1.png)

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Make prettier happy

Uli-Tiger committed:
fix: ensure ids used for name attributes and keys for cxFocus directive are unquie (#16727)

In case of immediate conflict resolution an attribute may get rendered twice, once in the conflict resolution popup and once still in background. As the id generation assumes that an attribute is only rendered once, this leads to duplicate ids. In most cases this does not cause any issues, however there are 2 situations where ids need to be unique.
*  key of cxFocus directive
*  name attributes of controls
for these cases there is an additional uiContextKey injected into the generated ids.

ChristophHi committed:
Merge and resolve conflicts

ChristophHi committed:
Fix tests

ChristophHi committed:
Always show config form

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

ChristophHi committed:
Add tests and fix coverage for one existing method

ChristophHi committed:
Add reducer tests

ChristophHi committed:
clean up and increase test coverage

Uli-Tiger committed:
test: cypress test of conflict solver dialog feature (#16734)

Uli-Tiger committed:
fix: ng100 when input fields are involved in conflicts

Uli-Tiger committed:
fix: make focus ids unique for all ui types

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

ChristophHi committed:
Adjust action and method names

ChristophHi committed:
ensure consistent naming in module

Uli-Tiger committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

Larisa-Staroverova committed:
Fix styling

Uli-Tiger committed:
fix: trap focus on conflict solver

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

Larisa-Staroverova committed:
Make prettier happy

ChristophHi committed:
Feature/gh 15991 simplify dialog (#16767)

* No need to check for OV routing

* Remove route check

* Refactor and add tests

* Fix sonar issue

Larisa-Staroverova committed:
Refactoring:conflict solver dialog launcher gets a conflict group instead of a list of conflict groups

ChristophHi committed:
Fix dialog html

Larisa-Staroverova committed:
Make sonar happy

Larisa-Staroverova committed:
Fix failing tests

Larisa-Staroverova committed:
Make static code checks happy

ChristophHi committed:
Rename form (#16771)

Larisa-Staroverova committed:
Rename styling files for configurator-group component

Uli-Tiger committed:
fix: ng100 if conflict involves attribute with add. value

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Fix a11y issues for AVC conflict solver

Uli-Tiger committed:
fix: ng100 caused by focus handling

Uli-Tiger committed:
fix: ng100 errors by detaching form view while conflict solver is open

Uli-Tiger committed:
remove unintended changes

Larisa-Staroverova committed:
Fix focus issue by closing conflict solver dialog

ChristophHi committed:
First steps navigation

ChristophHi committed:
Fix check for conflict solver dialog

ChristophHi committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

ChristophHi committed:
Merge branch 'feature/GH-15991' into feature/GH-15991-enrich-navigation

Uli-Tiger committed:
test: add tests for conflict resolver handling

Uli-Tiger committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

Uli-Tiger committed:
test: react on conflict solver

ChristophHi committed:
Fix navigation

ChristophHi committed:
Add tests

Uli-Tiger committed:
Revert "fix: changed heading level for correct document heading hierarchy(CXSPA-1067) (#16770)"

This reverts commit 260a6468b0cc56fb0c45325f92b5901948135c66.

Uli-Tiger committed:
Revert "fix: make focus ids unique for all ui types"

This reverts commit 468da50bcd8eccb39a0dc51578f474eb77214064.

Uli-Tiger committed:
Revert "fix: ng100 when input fields are involved in conflicts"

This reverts commit 93fd2f373b301383bc221c9413e7799cfba620b6.

Uli-Tiger committed:
Revert "fix: ensure ids used for name attributes and keys for cxFocus directive are unquie (#16727)"

This reverts commit 69beebf6a83a5cabb6ae110725d07ab802f17b97.

Uli-Tiger committed:
Revert "fix: ng100 if conflict involves attribute with add. value"

This reverts commit 047085b5c674e054583a57b7013fea9b6918a38a.

Uli-Tiger committed:
revert leftovers of id refactoring

Uli-Tiger committed:
revert leftover of id refactoring #2

Uli-Tiger committed:
fix ng303, ng304 and npe in ngOnDestroy console errors in test

ChristophHi committed:
Merge and resolve conflicts

ChristophHi committed:
Correct navigation and add tests

ChristophHi committed:
Feature/gh 15991 enrich navigation (#16780)

ChristophHi committed:
Get rid of changes not really needed

ChristophHi committed:
Merge branch 'feature/GH-15991-enrich-navigation' into feature/GH-15991

ChristophHi committed:
Update configurator-attribute-single-selection-base.component.spec.ts

ChristophHi committed:
Remove non-needed change from PR

ChristophHi committed:
Remove minor clean up not needed for this PR

Uli-Tiger committed:
Merge branch 'develop' into feature/GH-15991

Uli-Tiger committed:
Revert "Revert "fix: changed heading level for correct document heading hierarchy(CXSPA-1067) (#16770)""

This reverts commit 070d235c2ec53fb9d0f01ec396fc3f3feda4d013.

Uli-Tiger committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

Uli-Tiger committed:
test: ensure conflict dialog is scrolled properly

ChristophHi committed:
Re-introduce loading animation if dialog is present

ChristophHi committed:
Remove obsolete

ChristophHi committed:
delay dialog and simplify service and tests

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Uli-Tiger committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

Uli-Tiger committed:
test: fix failing cypress tests

steinsebastian committed:
Move SR alert to top of dialog

steinsebastian committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

steinsebastian commented on feature-libs/product-configurator/rulebased/components/form/configurator-form.component.ts:
```diff
@@ -45,120 +50,111 @@ export class ConfiguratorFormComponent implements OnInit {
-      group.groupType !== undefined &&
-      this.configuratorGroupsService.isConflictGroupType(group.groupType) &&
-      group.name !== ''
+  isNavigationToGroupEnabled(): Observable<boolean> {
```
Wouldn't be "isImmediateConflictResolutionEnabled" a better name? This method/information is not only used for checking whether group navigation is enabled but more generally whether the immediate conflict resolution is enabled.

Uli-Tiger committed:
remove unused change detector ref

Uli-Tiger committed:
remove unused import

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

ChristophHi committed:
Remove navigation from form and facade layer

* Set current group already in reducer

ChristophHi committed:
Add missing test

ChristophHi committed:
Update configurator-form.component.ts

ChristophHi committed:
Add test

Larisa-Staroverova committed:
Set focus at header within conflict solver dialog

Uli-Tiger committed:
fix: ensure focus is always on top if conflict solver opens

Uli-Tiger committed:
add test for focus reset

Larisa-Staroverova committed:
Fix conflict-solver dialog styling for mobile widgets

steinsebastian committed:
Move tabindex to title to improve SR behavior

steinsebastian committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

ChristophHi committed:
Simplify code

ChristophHi committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

ChristophHi committed:
Expose conflict solver dialog to public API

ChristophHi committed:
Rename methods

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/product_configurator/product-configurator-vc-conflict-dialog-e2e-spec-flaky.ts:
```diff
@@ -0,0 +1,84 @@
+context('Product Configuration', () => {
+  beforeEach(() => {
+    cy.visit('/');
+    configurationVc.registerConfigurationRoute();
```
@Uli-Tiger, what about `registerGetConfigurationRoute`?

Uli-Tiger replied:
As this method was not introduced with this PR and as it used within most cypress tests, i would suggest not to rename it, as i it would enlarge the PR and complicate merging later.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/product_configurator/product-configurator-vc-conflict-dialog-e2e-spec-flaky.ts:
```diff
@@ -0,0 +1,84 @@
+  beforeEach(() => {
+    cy.visit('/');
+    configurationVc.registerConfigurationRoute();
+    configurationVc.registerConfigurationUpdateRoute();
```
@Uli-Tiger, what about `registerUpdateConfigurationRoute`?

Uli-Tiger replied:
As this method was not introduced with this PR and as it used within most cypress tests, i would suggest not to rename it, as i it would enlarge the PR and complicate merging later.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/product_configurator/product-configurator-vc-conflict-dialog-e2e-spec-flaky.ts:
```diff
@@ -0,0 +1,84 @@
+  });
+
+  it('should follow the immediate conflict resolution process', () => {
+    configurationVc.clickOnNextBtnAndWait('Specification');
```
@Uli-Tiger, it is not clear for what will be waiting. I would suggest `clickOnNextBtn` method name.

Uli-Tiger replied:
there is already a clickOnNextBtn method within configuration-helper (non vc), which does not wait for the GET request to happen. This method waits for the GET request.
As this method was not introduced with this PR and as it used within most cypress tests, i would suggest not to rename it, as i it would enlarge the PR and complicate merging later.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/product_configurator/product-configurator-vc-conflict-dialog-e2e-spec-flaky.ts:
```diff
@@ -0,0 +1,84 @@
+
+  it('should follow the immediate conflict resolution process', () => {
+    configurationVc.clickOnNextBtnAndWait('Specification');
+    configurationVc.selectAttributeAndWait('CAMERA_VIEWFINDER', RB, 'E'); // CONFLICT 1
```
@Uli-Tiger, it is not clear for what will be waiting. I would suggest `selectValue` method name.

Uli-Tiger replied:
there is already a selectValue method within configuration-helper (non vc), which does not wait for the GET request to happen. This method waits for the GET request.
As this method was not introduced with this PR and as it used within most cypress tests, i would suggest not to rename it, as i it would enlarge the PR and complicate merging later.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+import * as configurationVc from './product-configurator-vc';
+
+/**
+ * verifies that the conflict solver dialog is currently opened
```
@Uli-Tiger, plz, write `Verifies` with capital letter.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+}
+
+/**
+ * verifies that the conflict solver dialog is currently closed
```
@Uli-Tiger, plz, write `Verifies` with capital letter.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+}
+
+/**
+ * verifies the content of the conflict solver dialog
```
@Uli-Tiger, plz, write `Verifies` with capital letter.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+
+/**
+ * verifies the content of the conflict solver dialog
+ * @param involvedAttributes expected involved attributes for the currently visible conflict
```
@Uli-Tiger, plz, instead `involvedAttributes` I would suggest `conflictAttributes` or `involvedInConflictAttributes`.

Uli-Tiger replied:
conflictingAttribute

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+/**
+ * verifies that the conflict solver dialog has a title
+ */
+export function checkHasTitle() {
```
@Uli-Tiger, I would suggest `checkTitle` as a method name

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+/**
+ * verifies that the conflict solver dialog has a close button
+ */
+export function checkHasCloseButton() {
```
@Uli-Tiger,I would suggest `checkCloseBtn` or `checkCloseBtnExist`

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+}
+
+/**
+ * verifies that the conflict solver dialog has a close button
```
@Uli-Tiger, plz, write `Verifies` with capital letter.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+}
+
+/**
+ * verifies that the conflict solver dialog has a title
```
@Uli-Tiger, plz, write `Verifies` with capital letter.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+}
+
+/**
+ * verifies that the conflict solver dialog displays the resole issues message
```
@Uli-Tiger, plz, write `Verifies` with capital letter.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+}
+
+/**
+ * verifies that the conflict solver dialog displays the conflict description
```
@Uli-Tiger, plz, write `Verifies` with capital letter.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+}
+
+/**
+ * verifies that the conflict solver dialog displays the given number of suggestions
```
@Uli-Tiger, plz, write `Verifies` with capital letter.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+}
+
+/**
+ * verifies if given attribute is involved in the conflict
```
@Uli-Tiger, plz, write `Verifies` with capital letter.
What about `Verifies if the given attribute is involved in the conflict`?

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+  );
+}
+/**
+ * closes the conflict solver dialog via the close button
```
@Uli-Tiger, plz, write `Verifies` with capital letter.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+/**
+ * verifies that the conflict solver dialog displays the resole issues message
+ */
+export function checkHasResolveMessage() {
```
@Uli-Tiger, I would suggest `checkResolveMsgDisplayed`

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+/**
+ * verifies that the conflict solver dialog displays the conflict description
+ */
+export function checkHasDescription() {
```
@Uli-Tiger, I would suggest `checkDiscriptionDisplayed`

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+ * verifies that the conflict solver dialog displays the given number of suggestions
+ * @param {number} number number of expected suggestions
+ */
+export function checkNumberOfSuggestions(number: number) {
```
@Uli-Tiger, could you rename a parameter? I wonder why it works. In Java it is not allowed to use the same name.
What about `amount`?

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+
+/**
+ * verifies if given attribute is involved in the conflict
+ * @param {string} attributeName name of attribute
```
@Uli-Tiger, what about `attribute`, `selectedValues`?

Uli-Tiger replied:
we expect the attribute name here, not the whole attribute

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+ * @param {string} valueName value name to select / value to set
+ */
+export function selectAttributeAndWait(
+  attributeName: string,
```
@Uli-Tiger, what about `attribute`, `selectedValues`?
Could you change the order of the parameters? In the above method `uiType` is at the end

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-conflict-dialog.ts:
```diff
@@ -0,0 +1,136 @@
+ * @param {uiType} uiType uiType o attribute
+ * @param {string} valueName value name to select / value to set
+ */
+export function selectAttributeAndWait(
```
@Uli-Tiger, what about `selectValue` as a method name? The attribute will not be selected ;)

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator-vc.ts:
```diff
@@ -296,32 +296,63 @@ export function clickOnViewInConfiguration(attribute: string): void {
   attribute: string,
-  linkName: string
+  linkName: string,
+  linkDisplayed: boolean
```
@Uli-Tiger, what about `isLinkDispalyed`?

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator.ts:
```diff
@@ -565,3 +573,14 @@ export function completeOrderProcess(productName: string): void {
+/**
+ * Clicks on 'Proceed to Checkout' on the product details page.
+ */
+export function clickExitConfigurationButton(): void {
```
@Uli-Tiger, what about `clickExitConfigurationBtn`?

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/helpers/product-configurator.ts:
```diff
@@ -565,3 +573,14 @@ export function completeOrderProcess(productName: string): void {
 }
+
+/**
+ * Clicks on 'Proceed to Checkout' on the product details page.
```
@Uli-Tiger, plz, adjust the comment to the method.

Larisa-Staroverova commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/product_configurator/product-configurator-vc-conflict-dialog-e2e-spec-flaky.ts:
```diff
@@ -0,0 +1,84 @@
+    conflictDialog.checkIsClosed();
+  });
+
+  it('should restrict navigation when immediate conflict resolving is active', () => {
```
@Uli-Tiger, what about `should diable a classical conflict navigation navigation when the immediate conflict resolving is active`?

Uli-Tiger replied:
Actually it tests more than that the classical conflict navigation is disabled. For example it checks that the conflict solver re-opens if the prev/next buttons are pressed.

Uli-Tiger committed:
review feedback

Uli-Tiger committed:
Merge branch 'feature/GH-15991' of https://github.com/SAP/spartacus into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

Larisa-Staroverova committed:
Merge branch 'develop' into feature/GH-15991

ChristophHi committed:
Merge branch 'develop' into feature/GH-15991
